cmd/cli/cli.go

```diff
@@ -216,64 +216,102 @@ func sourceWriteFunc(t targetiface.Target, ft failureiface.Failure, tr transform
 		if len(messagesToSend) > 0 {
-			err2 := retry.Exponential(5, time.Second, "failureTarget.WriteOversized", func() error {
-				res, err := ft.WriteOversized(t.MaximumAllowedMessageSizeBytes(), messagesToSend)
-				if err != nil {
-					return err
-				}
-				if len(res.Oversized) != 0 || len(res.Invalid) != 0 {
+
+			writeOversized := func() (*models.TargetWriteResult, error) {
+				result, err := ft.WriteOversized(t.MaximumAllowedMessageSizeBytes(), messagesToSend)
+				if len(result.Oversized) != 0 || len(result.Invalid) != 0 {
 					log.Fatal("Oversized message transformation resulted in new oversized / invalid messages")
 				}
 
-				o.TargetWriteOversized(res)
-				messagesToSend = res.Failed
+				o.TargetWriteOversized(result)
+				messagesToSend = result.Failed
+				return result, err
+			}
+
+			_, err := handleWrite(cfg, writeOversized)
+
+			if err != nil {
 				return err
-			})
-			if err2 != nil {
-				return err2
 			}
 		}
 
 		// Send invalid message buffer
-		messagesToSend = append(resCast.Invalid, transformed.Invalid...)
+		messagesToSend = append(writeResult.Invalid, transformed.Invalid...)
 		if len(messagesToSend) > 0 {
-			err3 := retry.Exponential(5, time.Second, "failureTarget.WriteInvalid", func() error {
-				res, err := ft.WriteInvalid(messagesToSend)
-				if err != nil {
-					return err
-				}
-				if len(res.Oversized) != 0 || len(res.Invalid) != 0 {
+			writeInvalid := func() (*models.TargetWriteResult, error) {
+				result, err := ft.WriteInvalid(messagesToSend)
+				if len(result.Oversized) != 0 || len(result.Invalid) != 0 {
 					log.Fatal("Invalid message transformation resulted in new invalid / oversized messages")
 				}
 
-				o.TargetWriteInvalid(res)
-				messagesToSend = res.Failed
+				o.TargetWriteInvalid(result)
+				messagesToSend = result.Failed
+				return result, err
+			}
+
+			_, err := handleWrite(cfg, writeInvalid)
+
+			if err != nil {
 				return err
-			})
-			if err3 != nil {
-				return err3
 			}
 		}
-
 		return nil
 	}
 }
 
+// Wrap each target write operation with 2 kinds of retries:
+// - setup errors: long delay, unlimited attempts, unhealthy state + alerts
+// - transient errors: short delay, limited attempts
+// If it's setup/transient error is decided based on a response returned by the target.
+func handleWrite(cfg *config.Config, write func() (*models.TargetWriteResult, error)) (*models.TargetWriteResult, error) {
+	retryOnlySetupErrors := retry.RetryIf(func(err error) bool {
+		_, isSetup := err.(models.SetupWriteError)
+		return isSetup
+	})
+
+	onSetupError := retry.OnRetry(func(attempt uint, err error) {
+		// Here we can set unhealthy status + send monitoring alerts in the future. Nothing happens here now.
+	})
+
+	//First try to handle error as setup...
+	writeResult, err := retry.DoWithData[*models.TargetWriteResult](
+		write,
+		retryOnlySetupErrors,
+		onSetupError,
+		retry.Delay(time.Duration(cfg.Data.Retry.Transient.Delay)*time.Second),
```

Comment:
Would it be possible/practical to configure an exponential backoff, should we wish to?

Reply:
If not explicitly set, this retry lib by [default](https://github.com/avast/retry-go/blob/01d7e0007ca266461dd395bc76ee7d14b2a73b63/retry.go#L234) already uses [exponential backoff with jitter](https://github.com/avast/retry-go/blob/01d7e0007ca266461dd395bc76ee7d14b2a73b63/options.go#L115)